Add stolonctl clusterdata write sub command

dineshba opened the pull request:
This PR fixes #577 

This PR will:

- make **clusterdata** to have read and write functionality

```sh
stolonctl clusterdata read <options>
stolonctl clusterdata write <options>
```
**write** can read cluser data from stdin as well from file. In this way we can combine both read and write operations using `|`

```sh
stolonctl clustedata read --cluster-name example --store-backend etcd | stolonctl clusterdata write --cluster-name example-copy --store-backend consul
```

dineshba commented:
It's ready for review @sgotti. Please have a look and provide your feedbacks.

sgotti commented on cmd/stolonctl/cmd/clusterdata.go:
```diff
@@ -15,32 +15,76 @@
 }
 
-var clusterdataOpts clusterdataOptions
+type Reader interface {
```
I don't understand why this interface and the below implementation is needed. I think that the functions `readClusterData`, `readClusterDataFromFile` and `readClusterDataFromStdIn` don't really need it since they receive the filename from the global config `readClusterdataOpts`.

Or is this just needed for the tests to mock the reader? If so isn't there a way to do this without "bloating" the code?

dineshba replied:
We introduced `Reader` interface only for **unit testing**. We added unit tests using the MockReader (which are generated using interface Reader). If you think, it is bloating the code, we can remove the interface and write integration tests. Thoughts?

We are using `readClusterDataFromFile`, `readClusterDataFromStdIn` to read the new cluster data to be written to the new store. We will read from file or stdin based on `writeClusterdataOpts.file`. So we need both the methods. Thoughts ?

dineshba replied:
Hi @sgotti. We **removed** `Reader` interface and methods `readClusterData`, `readClusterDataFromFile` and `readClusterDataFromStdIn`. We are using different way to achieve the unit testing.

Now we sending reader to `writeClusterdata` based on user input. For files, we are sending `File`(as File is a Reader) and `Stdin` (as Stdin is a Reader). For testing we are sending simple `StringReader`. So we able **easily test without mocking** Reader.

**Thanks for your inputs. It helped a lot and we removed lot of bloating code**. Please have a look and provide your inputs.

dineshba commented on cmd/stolonctl/cmd/clusterdata.go:
```diff
@@ -15,32 +15,76 @@
+	"context"
 	"encoding/json"
-
+	"fmt"
```
Sure. Changed.

dineshba commented:
Hi @sgotti, Added dependencies in go.mod also  (used latest go1.11.2).

dineshba commented on internal/mock/store/store.go:
```diff
@@ -0,0 +1,210 @@
+// Source: internal/store/store.go
+
+// Package mock_store is a generated GoMock package.
+package mock_store
```
Hi @sgotti. I added `go:generate` line like below in `store.go` file
```go
//go:generate mockgen -destination=../mock/store/store.go -source=$GOFILE
```
and when we do `go generate ./...` from our project directory, it is generating the file at the given destination. 

I am facing one **problem** there. While generating it is using local import for store like
```go
import (
	x "."
	context "context"
)
```
instead of 
```go
import (
	store "github.com/sorintlab/stolon/internal/store"
	context "context"
```

I spent some time to resolve this. But no luck. Please help me to fix this. In the mean time, I will also try to fix this. Thanks

dineshba commented:
Hi @sgotti 
1. I fixed the Semaphore k8s failing tests
2. Added the upgrade notes under section `v0.14.0`
3. Generated docs and pushed the latest doc files.

Please have a look and provide your inputs

rShetty commented on cmd/stolonctl/cmd/clusterdata.go:
```diff
@@ -67,3 +97,68 @@ func clusterdata(cmd *cobra.Command, args []string) {
+	}
+
+	if err = isSafeToWriteClusterData(s); err != nil {
+		return fmt.Errorf("%v", err)
```
Curious to know why can't we just return an `err` back here ?

dineshba replied:
@rShetty Thanks for pointing it out. It was by mistake. No particular reason.

dineshba commented:
Hi @sgotti Added warning in the docs [refer](https://github.com/jskswamy/stolon/blob/write-cluster-data/doc/commands/stolonctl_clusterdata_write.md). And squashed the commits.
Please have a look and provide your inputs

dineshba commented:
Hi @sgotti builds were failing for `checksum error` in go.sum file. Fixed it. Please have a look.
